### PR TITLE
[ADD] udes_stock: Add in "scanned by" functionality

### PR DESCRIPTION
**Pull request by caleb-shelton**
Add tracking data of who did the picking and the done datetime.
Add unit tests covering this.

User-story: 1083

Signed-off-by: Caleb Shelton <caleb.shelton@unipart.io>

### CODE REVIEW

File: addons/udes_stock/models/stock_move_line.py [hunk 1]
```diff
@@ -1,13 +1,33 @@
...
 
+_logger = logging.getLogger(__name__)
+
 
 class StockMoveLine(models.Model):
     _inherit = "stock.move.line"
 
+    u_done_by_id = fields.Many2one(
+        "res.users",
+        "Scanned by",
```
Comment (PeterC10) [inline addons/udes_stock/models/stock_move_line.py:18]: `Scanned By`

File: addons/udes_stock/models/stock_move_line.py [hunk 2]
```diff
@@ -1,13 +1,33 @@
...
+        help="ID of the user to complete the pack op.",
```
Comment (PeterC10) [inline addons/udes_stock/models/stock_move_line.py:19]: Suggestion: `User who completed the product move.`

File: addons/udes_stock/tests/test_scanned_by.py [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+        cls.create_quant(cls.banana.id, cls.test_stock_location_02.id, 20)
+        picking = cls.create_picking(cls.picking_type_internal)
+        move = cls.create_single_move(cls.apple, 10, picking)
+        move = cls.create_single_move(cls.banana, 20, picking)
+        picking.action_confirm()
+        picking.action_assign()
+
+        cls.picking = picking.with_env(test_user_env)
+
+    # TODO: Does this need enabling even with real_uid removed now?
```
Comment (PeterC10) [inline addons/udes_stock/tests/test_scanned_by.py:39]: Yes this is still needed, we still want to be sure that `sudo` calls do not change who completed the operation.

File: addons/udes_stock/models/stock_move_line.py (hunk 1 above)
Comment (caleb-shelton) [reply addons/udes_stock/models/stock_move_line.py:18]: I have https://capitalizemytitle.com/ book marked as well! :cry:

File: addons/udes_stock/models/stock_move_line.py [hunk 4]
```diff
@@ -16,6 +36,30 @@ class StockMoveLine(models.Model):
         readonly=True,
     )
 
+    @staticmethod
+    def _now_for_tracking_data():
+        """Can be overwritten to change the tracking resolution"""
```
Comment (kdwyer) [inline addons/udes_stock/models/stock_move_line.py:41]: I'd make this "Override to change ..."

File: addons/udes_stock/tests/test_scanned_by.py [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+        cls.create_quant(cls.banana.id, cls.test_stock_location_02.id, 20)
+        picking = cls.create_picking(cls.picking_type_internal)
+        move = cls.create_single_move(cls.apple, 10, picking)
+        move = cls.create_single_move(cls.banana, 20, picking)
+        picking.action_confirm()
+        picking.action_assign()
+
+        cls.picking = picking.with_env(test_user_env)
+
+    def test_records_real_id_on_move_line_update(self):
```
Comment (kdwyer) [inline addons/udes_stock/tests/test_scanned_by.py:39]: I'd name the method `test_records_original_user_id_on_move_line_update`, "real" doesn't seem to mean anything here.
In the docstring 
- u_done_by -> u_done_by_id?
- required -> used?
- and be reported on -> can be reported on?

File: addons/udes_stock/tests/test_scanned_by.py [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+        """The original user should be in u_done_by so that when sudo is required for
+        permission access the correct user and be reported on
+        """
+        sudo_picking = self.picking.sudo()  # Switch to admin to get around any access rights
+        move_line = sudo_picking.move_line_ids[0]
+        move_line.write({"qty_done": move_line.product_uom_qty})
+        self.assertEqual(move_line.u_done_by_id.id, self.test_user.id)
+
+    def test_multiple_users_move_line_update(self):
```
Comment (kdwyer) [inline addons/udes_stock/tests/test_scanned_by.py:48]: I think this method name could be a little more descriptive; maybe something like:
`test_records_correct_user_if_different_user_completes_picking`

(I'm not convinced about "correct_user" either - ideally we want to convey that it's the user who completes the picking.  Perhaps you can come up with something better)

File: addons/udes_stock/tests/test_scanned_by.py [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+        """Check that when two users act on the same picking that they are recorded seperately"""
```
Comment (kdwyer) [inline addons/udes_stock/tests/test_scanned_by.py:49]: > Check that ... they are recorded separately

We seem to be testing that the "second" user is recorded in `u_done_by_id` - the docstring should accurately reflect what we are testing.

File: addons/udes_stock/tests/test_scanned_by.py (hunk 6 above)
Comment (caleb-shelton) [reply addons/udes_stock/tests/test_scanned_by.py:48]: I think 'correct' works.

Comment (kdwyer) [reply addons/udes_stock/tests/test_scanned_by.py:48]: mmm, but the reader needs to understand what "correct" means in this context.  I'd rather find something more explicit, if we could.

File: addons/udes_stock/tests/test_scanned_by.py (hunk 5 above)
Comment (kdwyer) [reply addons/udes_stock/tests/test_scanned_by.py]: Apologies, an even better name would be

`test_records_original_user_id_on_move_line_update_with_sudo`

Comment (caleb-shelton) [reply addons/udes_stock/tests/test_scanned_by.py]: fixed

**Review by kdwyer — APPROVED** [review]

File: addons/udes_stock/tests/common.py [hunk 8]
```diff
@@ -391,6 +391,23 @@ def create_move(cls, pickings, products_info, **kwargs):
         moves = Picking._create_move(move_values)
         return moves
 
+    @classmethod
+    def create_single_move(cls, product, qty, picking, **kwargs):
```
Comment (peterpratt13) [inline addons/udes_stock/tests/common.py:395]: Q: Why do we do this when create_move exists?

File: addons/udes_stock/models/stock_move_line.py [hunk 9]
```diff
@@ -16,6 +36,30 @@ class StockMoveLine(models.Model):
...
+    def _add_user_tracking_data(self, vals):
+        """Inject/overwrite user tracking data into vals"""
+        vals["u_done_by_id"] = self.env.uid
+        vals["u_done_datetime"] = self._now_for_tracking_data()
+        return vals
+
+    @api.model
+    def create(self, vals):
+        """Extend to track the requester user"""
+        vals = self._add_user_tracking_data(vals)
```
Comment (peterpratt13) [inline addons/udes_stock/models/stock_move_line.py:54]: Does this mean the u_done_by on the move line is set on creation? When the qty_done is zero

File: addons/udes_stock/tests/common.py (hunk 8 above)
Comment (peterpratt13) [reply addons/udes_stock/tests/common.py:395]: Can we not just use create_move, which uses the functionality emplyed by the code outside of tests?

File: addons/udes_stock/models/stock_move_line.py (hunk 9 above)
Comment (caleb-shelton) [reply addons/udes_stock/models/stock_move_line.py:54]: Does this have a potential to break something?

File: addons/udes_stock/tests/common.py (hunk 8 above)
Comment (caleb-shelton) [reply addons/udes_stock/tests/common.py:395]: fixed

File: addons/udes_stock/models/stock_move_line.py (hunk 9 above)
Comment (peterpratt13) [reply addons/udes_stock/models/stock_move_line.py:54]: qty_done > 0 i believe

File: addons/udes_stock/tests/common.py [hunk 10]
```diff
@@ -391,6 +391,23 @@ def create_move(cls, pickings, products_info, **kwargs):
...
+            "name": product.name,
+            "product_uom": product.uom_id.id,
+            "product_uom_qty": qty,
+            "location_id": picking.location_id.id,
+            "location_dest_id": picking.location_dest_id.id,
+            "picking_id": picking.id,
+            "priority": picking.priority,
+        }
+        vals.update(kwargs)
+        return Move.create(vals)
```
Comment (peterpratt13) [inline addons/udes_stock/tests/common.py:409]: Can delete this now

File: addons/udes_stock/models/stock_move_line.py (hunk 9 above)
Comment (caleb-shelton) [reply addons/udes_stock/models/stock_move_line.py:55]: fixed

**Review by peterpratt13 — APPROVED** [review]

**Pull request merged**